### PR TITLE
scripts(bin/update-checksum): support for array SHA-256

### PR DESCRIPTION
**Pull request by tqfx**
See #20529

### CODE REVIEW

**Comment by tqfx** [comment]
See #20529

File: scripts/bin/update-checksum [hunk 1]
```diff
@@ -57,29 +57,35 @@ for package in "${@}"; do
...
-		fi
-	)
-
-	if [ -n "${new_checksum}" ]; then
+	if [ -n "${checksum}" ] && [ -n "${new_checksum}" ]; then
 		# Replace old SHA-256.
-		# Note that this will not work properly if SHA-256 is an array.
-		sed -i "s/^\(TERMUX_PKG_SHA256=\)\(.*\)\$/\1${new_checksum}/g" "${buildsh_path}"
+		for ((i=0; i<${#new_checksum[@]}; ++i)) do
+			sed -i "s|${checksum[i]}|${new_checksum[i]}|" "${buildsh_path}"
```
Comment (truboxl) [inline scripts/bin/update-checksum:87]: This breaks the old way of quickly populate empty or replace invalid `TERMUX_PKG_SHA256=` when prototyping `build.sh`

**Comment by tqfx** [comment]
Work with case 1

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=(https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
                   https://github.com/ArtalkJS/Artalk/releases/download/v${TERMUX_PKG_VERSION}/artalk_ui.tar.gz)
TERMUX_PKG_SHA256=
```

Work with case 2

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=(https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
                   https://github.com/ArtalkJS/Artalk/releases/download/v${TERMUX_PKG_VERSION}/artalk_ui.tar.gz)
TERMUX_PKG_SHA256=invalid
```

Work with case 3

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=(https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
                   https://github.com/ArtalkJS/Artalk/releases/download/v${TERMUX_PKG_VERSION}/artalk_ui.tar.gz)
TERMUX_PKG_SHA256=(
invalid
)
```

Work with case 4

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_SHA256=
```

Work with case 5

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_SHA256=invalid
```

Work with case 6

```sh
TERMUX_PKG_VERSION=2.8.7
TERMUX_PKG_SRCURL=https://github.com/ArtalkJS/Artalk/archive/v${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_SHA256=(
invalid
)
```

File: scripts/bin/update-checksum [hunk 2]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
-			rm -rf "${dl_tmpdir}"
-		fi
-	)
+		done
+	))
 
 	if [ -n "${new_checksum}" ]; then
+		if [ ${#new_checksum[@]} -gt 1 ]; then
+			checksum="(\n"
+			for ((i=0; i<${#new_checksum[@]}; ++i)) do
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:83]: We can turn this entire for loop into a single `printf`
e.g.
```suggestion
			checksum+="$( printf '\t%s\n' "${new_checksum[@]}" )"
```

File: scripts/bin/update-checksum [hunk 3]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
 			fi
-
-			rm -rf "${dl_tmpdir}"
-		fi
-	)
+		done
+	))
 
 	if [ -n "${new_checksum}" ]; then
+		if [ ${#new_checksum[@]} -gt 1 ]; then
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:81]:
```suggestion
		if (( ${#new_checksum[@]} > 1 )); then
```

File: scripts/bin/update-checksum [hunk 4]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
 
 	# Calculate new checksum in subshell.
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:58]: This comment could use an update.
```suggestion
	# Command substitution create a subshell, we want to do this 
	# to not effect the main scope's variables by sourcing the build script
```
Something along these lines, feel free to come up with something less wordy.

File: scripts/bin/update-checksum [hunk 5]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
 	echo
-	new_checksum=$(
+	new_checksum=($(
 		source "${buildsh_path}" 2>/dev/null
+		for url in ${TERMUX_PKG_SRCURL[@]}; do
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:62]: This should be quoted.
https://www.shellcheck.net/wiki/SC2068
```suggestion
		for url in "${TERMUX_PKG_SRCURL[@]}"; do
```

File: scripts/bin/update-checksum [hunk 6]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
+				rm -rf "${dl_tmpdir}"
 			fi
-
-			rm -rf "${dl_tmpdir}"
-		fi
-	)
+		done
+	))
 
 	if [ -n "${new_checksum}" ]; then
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:80]: That's an array now.
Also prefer using arithmetic evaluation for a null length check.
```suggestion
	if (( ${#new_checksum[@]} )); then
```

File: scripts/bin/update-checksum [hunk 7]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
 	# Calculate new checksum in subshell.
 	echo
-	new_checksum=$(
+	new_checksum=($(
 		source "${buildsh_path}" 2>/dev/null
+		for url in ${TERMUX_PKG_SRCURL[@]}; do
+			dl_tmpdir=$(mktemp -d "${TMPDIR-/tmp}/termux.src.dl.XXXXXXXX")
 
-		dl_tmpdir=$(mktemp -d "${TMPDIR-/tmp}/termux.src.dl.XXXXXXXX")
+			if [ -n "${dl_tmpdir}" ] && [ -d "${dl_tmpdir}" ]; then
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:65]: The `-n` test is redundant.
Also prefer `[[` bash test `]]`.
```suggestion
			if [[ -d "${dl_tmpdir}" ]]; then
```

File: scripts/bin/update-checksum [hunk 8]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
+					"$url"; then
+					rm -rf "${dl_tmpdir}"
+				fi
 
-		if [ -n "${dl_tmpdir}" ] && [ -d "${dl_tmpdir}" ]; then
-			if ! curl --fail --location --retry 3 --output "${dl_tmpdir}/source-bundle" \
-				"${TERMUX_PKG_SRCURL[0]}"; then
-				rm -rf "${dl_tmpdir}"
-			fi
+				if [ -f "${dl_tmpdir}/source-bundle" ]; then
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:71]: Prefer `[[` bash test `]]`.
```suggestion
				if [[ -f "${dl_tmpdir}/source-bundle" ]]; then
```

File: scripts/bin/update-checksum [hunk 9]
```diff
@@ -57,29 +57,42 @@ for package in "${@}"; do
...
 	echo
-	new_checksum=$(
+	new_checksum=($(
 		source "${buildsh_path}" 2>/dev/null
+		for url in ${TERMUX_PKG_SRCURL[@]}; do
+			dl_tmpdir=$(mktemp -d "${TMPDIR-/tmp}/termux.src.dl.XXXXXXXX")
 
-		dl_tmpdir=$(mktemp -d "${TMPDIR-/tmp}/termux.src.dl.XXXXXXXX")
+			if [ -n "${dl_tmpdir}" ] && [ -d "${dl_tmpdir}" ]; then
+				if ! curl --fail --location --retry 3 --output "${dl_tmpdir}/source-bundle" \
```
Comment (TomJo2000) [inline scripts/bin/update-checksum:66]: <sub>(Nitpick, feel free to ignore)</sub>
Not a big fan of this being split here.
```suggestion
				if ! curl --fail --location --retry 3 --output "${dl_tmpdir}/source-bundle" "$url"; then
```
Either keep it on 1 line, or split it after `--retry 3`.

**Review by TomJo2000 — CHANGES_REQUESTED** [review]
Just found this PR now.
There's a bit of clean up I'd like to do on this,
other than my couple nitpicks, great work.

**Comment by truboxl** [comment]
Breaks for opencv. Probably that build.sh need to be rewritten.

**Comment by TomJo2000** [comment]
> Breaks for opencv. Probably that build.sh need to be rewritten.

Yep, at least the srcurl and sha256.

**Comment by TomJo2000** [comment]
looks like we have a couple more of these.
A cursory `rg -Hni '\+=\(' {x11-,root-,}packages` points out:
- `x11-packages/opencv`
<sup> already handled</sup>
- `x11-packages/virglrenderer-android/`
<sup> already handled</sup>
- `packages/unar`
- `packages/libmusicbrainz`
- `packages/softethervpn`
- `packages/asymptote`
- `packages/maxima`
- `packages/alist`
- `packages/ldc`

**Comment by TomJo2000** [comment]
Wait a minute, do we not care about this for `TERMUX_PKG_VERSION`?
I think if we're "fixing" it for `TERMUX_PKG_SRCURL` and `TERMUX_PKG_SHA256` we should phase out `+=(` for `TERMUX_PKG_VERSION` as well.

**Comment by tqfx** [comment]
Okay. I'm going to modify them.

**Comment by tqfx** [comment]
If `TERMUX_PKG_VERSION` is a string, auto-updates will work. If `TERMUX_PKG_VERSION` is an array, things get a little more complicated. See [ldc](https://github.com/termux/termux-packages/blob/master/packages/ldc/build.sh). For this case I tend to override `termux_pkg_auto_update` in `build.sh`.

**Comment by TomJo2000** [comment]
> If `TERMUX_PKG_VERSION` is a string, auto-updates will work. If `TERMUX_PKG_VERSION` is an array, things get a little more complicated. See [ldc](https://github.com/termux/termux-packages/blob/master/packages/ldc/build.sh). For this case I tend to override `termux_pkg_auto_update` in `build.sh`.

Good point.
A version array would complicate auto-updates.

File: disabled-packages/mingw-w64-crt/build.sh
Comment (TomJo2000) [inline disabled-packages/mingw-w64-crt/build.sh:1]: No need to update disabled packages.
I won't comment on the rest of them.

File: packages/aapt/build.sh [hunk 10]
```diff
@@ -6,23 +6,27 @@ _TAG_VERSION=13.0.0
...
-TERMUX_PKG_SRCURL=(https://android.googlesource.com/platform/frameworks/base
-                   https://android.googlesource.com/platform/system/core
-                   https://android.googlesource.com/platform/system/libbase
-                   https://android.googlesource.com/platform/system/libziparchive
-                   https://android.googlesource.com/platform/system/logging
-                   https://android.googlesource.com/platform/system/incremental_delivery
-                   https://android.googlesource.com/platform/build
-                   https://android.googlesource.com/platform/system/tools/aidl)
+TERMUX_PKG_SRCURL=(
+	https://android.googlesource.com/platform/frameworks/base
```
Comment (TomJo2000) [inline packages/aapt/build.sh:10]: I guess not having the starting or closing parenthesis on the same line makes automatically updating it a lot easier.

If we wanna start enforcing that, that actually lets me simplify the package linter a bit.

File: packages/artalk/build.sh [hunk 11]
```diff
@@ -3,48 +3,17 @@ TERMUX_PKG_DESCRIPTION="A self-hosted comment system"
...
+	https://github.com/ArtalkJS/Artalk/releases/download/v${TERMUX_PKG_VERSION}/artalk_ui.tar.gz
+)
+TERMUX_PKG_SHA256=(
+	62e6fb2dbd8fa5c93a639b83874dafbe1ea3893477760a9e764a8085cdbc94e1
+	619b2d823cfaab7d9e98a4520b9f3f4d1f0e61d4aaf6022b13fb12f1830540ba
+)
 TERMUX_PKG_BUILD_IN_SRC=true
 TERMUX_PKG_AUTO_UPDATE=true
 
-termux_pkg_auto_update() {
```
Comment (TomJo2000) [inline packages/artalk/build.sh:13]: So this is able to be handled by the standard auto-update function now?

File: packages/asymptote/build.sh [hunk 12]
```diff
@@ -2,12 +2,18 @@ TERMUX_PKG_HOMEPAGE=https://asymptote.sourceforge.io/
...
 TERMUX_PKG_LICENSE="GPL-3.0"
 TERMUX_PKG_MAINTAINER="@termux"
-TERMUX_PKG_VERSION=(2.89)
-TERMUX_PKG_VERSION+=(1.0.1)
-TERMUX_PKG_SRCURL=(https://downloads.sourceforge.net/asymptote/asymptote-${TERMUX_PKG_VERSION}.src.tgz
-                   https://github.com/g-truc/glm/archive/${TERMUX_PKG_VERSION[1]}.tar.gz)
-TERMUX_PKG_SHA256=(f64e62b4ee4f85f1a78640c4f1e8a6f98e91f54edacab19727c7cabe94a57f5b
-                   9f3174561fd26904b23f0db5e560971cbf9b3cbda0b280f04d5c379d03bf234c)
+TERMUX_PKG_VERSION=(
+	2.89
```
Comment (TomJo2000) [inline packages/asymptote/build.sh:6]: Comments to note which version is for what would be nice like some other packages have.

File: packages/asymptote/build.sh [hunk 13]
```diff
@@ -2,12 +2,18 @@ TERMUX_PKG_HOMEPAGE=https://asymptote.sourceforge.io/
...
+	1.0.1
+)
+TERMUX_PKG_SRCURL=(
+	https://downloads.sourceforge.net/asymptote/asymptote-${TERMUX_PKG_VERSION}.src.tgz
```
Comment (TomJo2000) [inline packages/asymptote/build.sh:10]: Explicitly use index `[0]` to show that it's an array.
<sup>Also I'm pretty sure Shellcheck yells at you for not indexing on arrays</sup>
```suggestion
	https://downloads.sourceforge.net/asymptote/asymptote-${TERMUX_PKG_VERSION[0]}.src.tgz
```

File: packages/groovy/build.sh [hunk 14]
```diff
@@ -8,13 +8,17 @@ _JANSI_VERSION=2.4.1
 _JLINE_VERSION=2.14.6
```
Comment (TomJo2000) [inline packages/groovy/build.sh:8]: If we're doing it for all these other packages,
we may as well fold these into a `TERMUX_PKG_VERSION=(...)` array.
The `_JANSI_TAGNAME` and `_JLINE2_TAGNAME` abstractions are probably still useful to keep though.

File: packages/ldc/build.sh [hunk 15]
```diff
@@ -88,7 +93,7 @@ termux_step_pre_configure() {
 		# [...]/ldc/src/llvm/projects/compiler-rt/lib/builtins/clear_cache.c:85:20:
 		# error: write to reserved register 'R7'
 		#   __asm __volatile("svc 0x0"
-		#                    ^
+		#	 ^
```
Comment (TomJo2000) [inline packages/ldc/build.sh:96]: I don't think that should have been modified.

File: packages/libmusicbrainz/build.sh [hunk 16]
```diff
@@ -2,17 +2,22 @@ TERMUX_PKG_HOMEPAGE=https://musicbrainz.org/doc/libmusicbrainz
...
-TERMUX_PKG_VERSION+=(2.9.12) # libxml2 version
-TERMUX_PKG_VERSION+=(0.32.1) # libneon version
-TERMUX_PKG_SRCURL=(https://github.com/metabrainz/libmusicbrainz/releases/download/release-${TERMUX_PKG_VERSION}/libmusicbrainz-${TERMUX_PKG_VERSION}.tar.gz
-                   ftp://xmlsoft.org/libxml2/libxml2-${TERMUX_PKG_VERSION[1]}.tar.gz
-                   https://notroj.github.io/neon/neon-${TERMUX_PKG_VERSION[2]}.tar.gz)
-TERMUX_PKG_SHA256=(6749259e89bbb273f3f5ad7acdffb7c47a2cf8fcaeab4c4695484cef5f4c6b46
-                   c8d6681e38c56f172892c85ddc0852e1fd4b53b4209e7f4ebf17f7e2eae71d92
-                   05c54bc115030c89e463a4fb28d3a3f8215879528ba5ca70d676d3d21bf3af52)
+TERMUX_PKG_SRCURL=(
+	https://github.com/metabrainz/libmusicbrainz/releases/download/release-${TERMUX_PKG_VERSION}/libmusicbrainz-${TERMUX_PKG_VERSION}.tar.gz
```
Comment (TomJo2000) [inline packages/libmusicbrainz/build.sh:12]: I can hear Shellcheck yelling at you.
```suggestion
	https://github.com/metabrainz/libmusicbrainz/releases/download/release-${TERMUX_PKG_VERSION[0]}/libmusicbrainz-${TERMUX_PKG_VERSION[0]}.tar.gz
```

File: packages/maxima/build.sh [hunk 17]
```diff
@@ -2,14 +2,19 @@ TERMUX_PKG_HOMEPAGE="https://maxima.sourceforge.io/"
...
+	5.47.0
+	24.5.10
+)
 TERMUX_PKG_REVISION=2
-TERMUX_PKG_SRCURL=(https://downloads.sourceforge.net/sourceforge/maxima/Maxima-source/$TERMUX_PKG_VERSION-source/maxima-$TERMUX_PKG_VERSION.tar.gz
-                   https://common-lisp.net/project/ecl/static/files/release/ecl-${TERMUX_PKG_VERSION[1]}.tgz)
-TERMUX_PKG_SHA256=(9104021b24fd53e8c03a983509cb42e937a925e8c0c85c335d7709a14fd40f7a
-                   e4ea65bb1861e0e495386bfa8bc673bd014e96d3cf9d91e9038f91435cbe622b)
+TERMUX_PKG_SRCURL=(
+	https://downloads.sourceforge.net/sourceforge/maxima/Maxima-source/$TERMUX_PKG_VERSION-source/maxima-$TERMUX_PKG_VERSION.tar.gz
```
Comment (TomJo2000) [inline packages/maxima/build.sh:11]:
```suggestion
	https://downloads.sourceforge.net/sourceforge/maxima/Maxima-source/${TERMUX_PKG_VERSION[0]}-source/maxima-${TERMUX_PKG_VERSION[0]}.tar.gz
```

File: packages/maxima/build.sh [hunk 18]
```diff
@@ -47,7 +52,7 @@ termux_step_host_build() {
 
 termux_step_make() {
 	local _PREFIX_FOR_BUILD=$TERMUX_PKG_HOSTBUILD_DIR/prefix
-	
+
```
Comment (TomJo2000) [inline packages/maxima/build.sh:55]: It looks like your formatter is taking out some tabs on occasion.
In this case, that's good though.
I'm not sure how this ever passed the trailing whitespace check.

File: packages/ncurses/build.sh [hunk 19]
```diff
@@ -9,22 +9,28 @@ TERMUX_PKG_MAINTAINER="@termux"
 # Using this simplifies things (no need to avoid downloading and applying patches manually),
 # and uses github is a high available hosting.
 _SNAPSHOT_COMMIT=8bd5a3d98fc741bdcc9e5fada1a3d5980e1ea22a
-TERMUX_PKG_VERSION=(6.4.20231001
-                    9.31
-                    15
-                    0.31.0
-                    0.11.0)
+TERMUX_PKG_VERSION=(
+	6.4.20231001
```
Comment (TomJo2000) [inline packages/ncurses/build.sh:13]: Comments would be nice here while we're at it.

File: packages/perl/build.sh [hunk 20]
```diff
@@ -9,14 +9,20 @@ TERMUX_PKG_MAINTAINER="@termux"
 # - libregexp-assemble-perl
 # - psutils
 # - subversion
-TERMUX_PKG_VERSION=(5.38.2
-                    388d0eedbfc3864bbbf7ad7f965064d99cac5aaa)
+TERMUX_PKG_VERSION=(
+	5.38.2
+	388d0eedbfc3864bbbf7ad7f965064d99cac5aaa
```
Comment (TomJo2000) [inline packages/perl/build.sh:14]: This one could use comments.
So, is that a SVN commit, or?

File: packages/perl/build.sh [hunk 21]
```diff
@@ -9,14 +9,20 @@ TERMUX_PKG_MAINTAINER="@termux"
...
-                   a975c196075623f0dc94f57d00633b0d18ed08e3d85a3ea19d34ece4ec1a94c1)
-TERMUX_PKG_SRCURL=(http://www.cpan.org/src/5.0/perl-${TERMUX_PKG_VERSION}.tar.gz
-                   https://github.com/arsv/perl-cross/archive/${TERMUX_PKG_VERSION[1]}.tar.gz)
-#                  https://github.com/arsv/perl-cross/releases/download/${TERMUX_PKG_VERSION[1]}/perl-cross-${TERMUX_PKG_VERSION[1]}.tar.gz)
+TERMUX_PKG_SHA256=(
+	a0a31534451eb7b83c7d6594a497543a54d488bc90ca00f5e34762577f40655e
+	a975c196075623f0dc94f57d00633b0d18ed08e3d85a3ea19d34ece4ec1a94c1
+)
+TERMUX_PKG_SRCURL=(
+	http://www.cpan.org/src/5.0/perl-${TERMUX_PKG_VERSION}.tar.gz
```
Comment (TomJo2000) [inline packages/perl/build.sh:22]:
```suggestion
	http://www.cpan.org/src/5.0/perl-${TERMUX_PKG_VERSION[0]}.tar.gz
```

File: packages/perl/build.sh [hunk 22]
```diff
@@ -9,14 +9,20 @@ TERMUX_PKG_MAINTAINER="@termux"
...
+	https://github.com/arsv/perl-cross/archive/${TERMUX_PKG_VERSION[1]}.tar.gz
+#	https://github.com/arsv/perl-cross/releases/download/${TERMUX_PKG_VERSION[1]}/perl-cross-${TERMUX_PKG_VERSION[1]}.tar.gz
```
Comment (TomJo2000) [inline packages/perl/build.sh:24]:
```suggestion

```

File: packages/rw/build.sh [hunk 23]
```diff
@@ -3,8 +3,8 @@ TERMUX_PKG_DESCRIPTION="A Rust implementation of sponge(1) that never write to T
 TERMUX_PKG_LICENSE="MIT"
 TERMUX_PKG_MAINTAINER="@termux"
 TERMUX_PKG_VERSION=1.0.1
-TERMUX_PKG_SRCURL=("https://github.com/jridgewell/rw/archive/c13c24e011ef5a79ea60bc51bb0d3fa930326146.tar.gz")
-TERMUX_PKG_SHA256=(699c32045c713bcfc8e7b89d5bd24d89d1cbb887ba8570b857391f98b64e6a9a)
+TERMUX_PKG_SRCURL=https://github.com/jridgewell/rw/archive/c13c24e011ef5a79ea60bc51bb0d3fa930326146.tar.gz
```
Comment (TomJo2000) [inline packages/rw/build.sh:6]: > [!IMPORTANT]
> Why are we hardcoding the commit here?
> I think you can drop this one and I'll fix it in a separate PR.

File: packages/softether-vpn/build.sh [hunk 24]
```diff
@@ -2,13 +2,19 @@ TERMUX_PKG_HOMEPAGE=https://www.softether.org/
...
+	1.0.18
+)
 TERMUX_PKG_REVISION=4
-TERMUX_PKG_VERSION+=(1.0.18)
-TERMUX_PKG_SRCURL=(https://github.com/SoftEtherVPN/SoftEtherVPN/releases/download/${TERMUX_PKG_VERSION}/SoftEtherVPN-${TERMUX_PKG_VERSION}.tar.xz
-                   https://github.com/jedisct1/libsodium/archive/${TERMUX_PKG_VERSION[1]}-RELEASE.tar.gz)
-TERMUX_PKG_SHA256=(b5649a8ea3cc6477325e09e2248ef708d434ee3b2251eb8764bcfc15fb1de456
-                   b7292dd1da67a049c8e78415cd498ec138d194cfdb302e716b08d26b80fecc10)
+TERMUX_PKG_SRCURL=(
+	https://github.com/SoftEtherVPN/SoftEtherVPN/releases/download/${TERMUX_PKG_VERSION}/SoftEtherVPN-${TERMUX_PKG_VERSION}.tar.xz
```
Comment (TomJo2000) [inline packages/softether-vpn/build.sh:11]:
```suggestion
	https://github.com/SoftEtherVPN/SoftEtherVPN/releases/download/${TERMUX_PKG_VERSION[0]}/SoftEtherVPN-${TERMUX_PKG_VERSION[0]}.tar.xz
```

File: packages/texinfo/build.sh
Comment (TomJo2000) [inline packages/texinfo/build.sh:1]: These parts of the custom auto_update function are being superseded by the standard behavior of `termux_pkg_upgrade_version`?

File: packages/unar/build.sh [hunk 25]
```diff
@@ -2,14 +2,19 @@ TERMUX_PKG_HOMEPAGE=https://theunarchiver.com/command-line
...
+)
 TERMUX_PKG_REVISION=8
-TERMUX_PKG_VERSION+=(1.10.7)
-TERMUX_PKG_VERSION+=(1.1)
-TERMUX_PKG_SRCURL=(https://github.com/MacPaw/XADMaster/archive/v${TERMUX_PKG_VERSION}/XADMaster-${TERMUX_PKG_VERSION}.tar.gz
-                   https://github.com/MacPaw/universal-detector/archive/${TERMUX_PKG_VERSION[1]}/universal-detector-${TERMUX_PKG_VERSION[1]}.tar.gz)
-TERMUX_PKG_SHA256=(3d766dc1856d04a8fb6de9942a6220d754d0fa7eae635d5287e7b1cf794c4f45
-                   8e8532111d0163628eb828a60d67b53133afad3f710b1967e69d3b8eee28a811)
+TERMUX_PKG_SRCURL=(
+	https://github.com/MacPaw/XADMaster/archive/v${TERMUX_PKG_VERSION}/XADMaster-${TERMUX_PKG_VERSION}.tar.gz
```
Comment (TomJo2000) [inline packages/unar/build.sh:11]: You know the drill.
```suggestion
	https://github.com/MacPaw/XADMaster/archive/v${TERMUX_PKG_VERSION[0]}/XADMaster-${TERMUX_PKG_VERSION[0]}.tar.gz
```

File: scripts/bin/update-checksum
Comment (TomJo2000) [inline scripts/bin/update-checksum:1]: I'll review this again in a separate pass later.
But I think we've covered everything in here already.

File: x11-packages/tigervnc/build.sh [hunk 26]
```diff
@@ -3,13 +3,19 @@ TERMUX_PKG_DESCRIPTION="Suite of VNC servers. Based on the VNC 4 branch of Tight
...
+	1.10.1
+	21.1.8
+)
 TERMUX_PKG_REVISION=40
-TERMUX_PKG_SRCURL=(https://github.com/TigerVNC/tigervnc/archive/v${TERMUX_PKG_VERSION}.tar.gz
-                   https://xorg.freedesktop.org/releases/individual/xserver/xorg-server-${TERMUX_PKG_VERSION[1]}.tar.xz)
-TERMUX_PKG_SHA256=(19fcc80d7d35dd58115262e53cac87d8903180261d94c2a6b0c19224f50b58c4
-                   38aadb735650c8024ee25211c190bf8aad844c5f59632761ab1ef4c4d5aeb152)
+TERMUX_PKG_SRCURL=(
+	https://github.com/TigerVNC/tigervnc/archive/v${TERMUX_PKG_VERSION}.tar.gz
```
Comment (TomJo2000) [inline x11-packages/tigervnc/build.sh:12]:
```suggestion
	https://github.com/TigerVNC/tigervnc/archive/v${TERMUX_PKG_VERSION[0]}.tar.gz
```

File: packages/ldc/build.sh [hunk 27]
```diff
@@ -2,23 +2,28 @@ TERMUX_PKG_HOMEPAGE=https://github.com/ldc-developers/ldc
...
-                   840cd65bf5f0dd06ca688f63b94d71fccd92b526bbf1d3892fe5535b1e85c10e
-                   5784d4cc47d0845af0897d3b7473a08dd0281a4cdabac0a486740840d014fde1)
+TERMUX_PKG_VERSION=(
+	1.30.0
+	14.0.3  # LLVM version
+	2.100.1 # TOOLS version
+	1.30.0  # DUB version
+)
+TERMUX_PKG_SRCURL=(
+	https://github.com/ldc-developers/ldc/releases/download/v${TERMUX_PKG_VERSION}/ldc-${TERMUX_PKG_VERSION}-src.tar.gz
```
Comment (TomJo2000) [inline packages/ldc/build.sh:12]:
```suggestion
	https://github.com/ldc-developers/ldc/releases/download/v${TERMUX_PKG_VERSION[0]}/ldc-${TERMUX_PKG_VERSION[0]}-src.tar.gz
```

**Review by TomJo2000 — CHANGES_REQUESTED** [review]
Sorry for spamming your email.
I took the opportunity to turn this review into generalized cleanup for these packages.

**Comment by tqfx** [comment]
For multiple versions, I think it can be managed in the following form.

```sh
TERMUX_PKG_VERSION=""
_OTHER_VERSION=""
```

Then rewrite `termux_pkg_auto_update` in `build.sh`. This way auto-update works and `_OTHER_VERSION` is easily modified by `sed`.

**Comment by TomJo2000** [comment]
> For multiple versions, I think it can be managed in the following form.
> 
> ```shell
> TERMUX_PKG_VERSION=""
> _OTHER_VERSION=""
> ```
> 
> Then rewrite `termux_pkg_auto_update` in `build.sh`. This way auto-update works and `_OTHER_VERSION` is easily modified by `sed`.

We could use associative arrays.
```bash
declare -A TERMUX_PKG_VERSION=(
    [foo]='0.1.2'
    [bar]='6.4.12'
    [baz]='11.0-beta7'
)
```
But that wouldn't really be any less complicated.
I think an indexed array approach can work just fine,
but I'll need to do some testing to work something out.

**Comment by tqfx** [comment]
Associative arrays do not work with `termux_pkg_upgrade_version`.

**Comment by TomJo2000** [comment]
> Associative arrays do not work with `termux_pkg_upgrade_version`.

Would be extra work to make them.
I'll see about getting indexed arrays working for the version tomorrow.

**Comment by truboxl** [comment]
This PR is about update-checksum and interaction with build.sh. Anything else (changing termux_pkg_auto_update) is just adding noise at this point.

**Review by truboxl — APPROVED** [review]

File: x11-packages/virglrenderer-android/build.sh [hunk 28]
```diff
@@ -2,17 +2,41 @@ TERMUX_PKG_HOMEPAGE=https://virgil3d.github.io/
...
-TERMUX_PKG_SHA256+=(a7ced37f4102b745ac86d6a70a9da399cc139ff168ba6b8002b4d8d43c900c15)
+TERMUX_PKG_SRCURL=(
+	https://gitlab.freedesktop.org/virgl/virglrenderer/-/archive/virglrenderer-${TERMUX_PKG_VERSION}/virglrenderer-virglrenderer-${TERMUX_PKG_VERSION}.tar.gz
+	https://github.com/anholt/libepoxy/archive/refs/tags/${_LIBEPOXY_VERSION}.tar.gz
+)
+TERMUX_PKG_SHA256=(
+	446ab3e265f574ec598e77bdfbf0616ee3c77361f0574bec733ba4bac4df730a
+	a7ced37f4102b745ac86d6a70a9da399cc139ff168ba6b8002b4d8d43c900c15
+)
+TERMUX_PKG_AUTO_UPDATE=true
```
Comment (licy183) [inline x11-packages/virglrenderer-android/build.sh:16]: IMO `virglrender-android` should not be auto-updated. It needs tests before updating.

**Review by TomJo2000 — APPROVED** [review]

**Review by licy183 — APPROVED** [review]

**Pull request merged**